File: 210_Course_Schedule_II.py
class Solution:
    def buildGraph(self, n, pre):
        graph = [[] for _ in range(n)]
        for p in pre:
            graph[p[1]].append(p[0])
        return graph
    
    def computeIndegree(self, n, graph):
        degree = [0] * n
        for g in graph:
            for v in g:
                degree[v] += 1
        return degree
    
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = self.buildGraph(numCourses, prerequisites)
        degree = self.computeIndegree(numCourses, graph)
        res = []
        for i in range(numCourses):
            target = 0
            while target < numCourses:
                if not degree[target]:
                    break
                target += 1
            if target == numCourses:
                return []
            degree[target] -= 1
            res.append(target)
            for p in graph[target]:
                degree[p] -= 1
        return res

File: 210_Course_Schedule_II.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph (numCourses, vector<int> ());
        vector<int> indegree (numCourses, 0);
        queue<int> q;
        vector<int> res;
        
        for (auto pre : prerequisites) {
            graph[pre[1]].push_back(pre[0]);
            ++indegree[pre[0]];
        }

        for (int i = 0; i < numCourses; ++i) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        while (!q.empty()) {
            int target = q.front(); q.pop();
            res.push_back(target);
            for (auto adj : graph[target]) {
                if (--indegree[adj] == 0) {
                    q.push(adj);
                }
            }
        }

        return res.size() == numCourses ? res : vector<int> ();
    }
};
